drake/systems/frames/CoordinateFrame.h

```diff
@@ -0,0 +1,134 @@
+#ifndef __CoordinateFrame_H__
+#define __CoordinateFrame_H__
+
+#include <iostream>
+#include <string>
+#include <memory>
+#include <vector>
+#include <stdexcept>
+#include <initializer_list>
+
+#undef DLLEXPORT
+#if defined(WIN32) || defined(WIN64)
+#if defined(drakeSystem_EXPORTS)
+#define DLLEXPORT __declspec( dllexport )
+#else
+#define DLLEXPORT __declspec( dllimport )
+#endif
+#else
+#define DLLEXPORT
+#endif
+
+class DrakeSystem;
+typedef std::shared_ptr<DrakeSystem> DrakeSystemPtr;
+
+/// Every input, state, and output in a DynamicalSystem has a coordinate frame
+/// attached to it.  Many bugs can be avoided by forcing developers to be
+/// explicit about these coordinate systems when they make combinations of systems.
+
+class DLLEXPORT CoordinateFrame {
+public:
+  CoordinateFrame(const std::string& _name, const std::vector<std::string>& _coordinates)
+          : name(_name), coordinates(_coordinates) {};
+  CoordinateFrame(const std::string& _name, const unsigned int dim, const std::string& prefix)
+          : name(_name) {
+    for (int i=0; i<dim; i++) coordinates.push_back(prefix+std::to_string(i));
+  }
+  CoordinateFrame(const std::string& _name) : name(_name) {}
+
+  virtual ~CoordinateFrame(void) {  };
+
+  virtual unsigned int getDim() const { return coordinates.size(); };
+  virtual const std::string& getCoordinateName(unsigned int i) const {
+    if (i>=coordinates.size())
+      throw std::runtime_error("index exceeds dimension of coordinate frame");
+    return coordinates[i];
+  }
+  virtual std::vector<std::string> getCoordinateNames() const { return coordinates; }
+  virtual void setCoordinateNames(const std::vector<std::string>& _coordinates)
+  {
+    if (_coordinates.size()!=coordinates.size())
+      throw std::runtime_error("(CoordinateFrame:setCoordinateNames) this would change the dimension of the frame");  // which would, e.g. confuse a multi-coordinate frame that was pointing to this
+    coordinates = _coordinates;
+  }
+
+  virtual std::ostream& print(std::ostream& os) const {
+    os << "Coordinate Frame: " << name << " (" << getDim() << " elements)" << std::endl;
+
+    for (auto c : coordinates) {
+      os << "  " << c << std::endl;
+    }
+    return os;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const CoordinateFrame& m)
+  {
+    return m.print(os);
+  }
+
+  virtual DrakeSystemPtr setupLCMInputs(const DrakeSystemPtr& sys) { return sys; }
+  virtual DrakeSystemPtr setupLCMOutputs(const DrakeSystemPtr& sys) { return sys; }
```

Comment:
that doesn’t work.  trust me — it was tricky to avoid dynamic casts and #defines and keep the lcm dependency from being infectious.

> On Aug 22, 2015, at 9:56 PM, Twan Koolen <notifications@github.com> wrote:
> 
> In drake/systems/frames/CoordinateFrame.h <https://github.com/RobotLocomotion/drake/pull/1274#discussion_r37701816>:
> 
> > +  virtual std::ostream& print(std::ostream& os) const {
> > +    os << "Coordinate Frame: " << name << " (" << getDim() << " elements)" << std::endl;
> > +
> > +    for (auto c : coordinates) {
> > +      os << "  " << c << std::endl;
> > +    }
> > +    return os;
> > +  }
> > +
> > +  friend std::ostream& operator<<(std::ostream& os, const CoordinateFrame& m)
> > +  {
> > +    return m.print(os);
> > +  }
> > +
> > +  virtual DrakeSystemPtr setupLCMInputs(const DrakeSystemPtr& sys) { return sys; }
> > +  virtual DrakeSystemPtr setupLCMOutputs(const DrakeSystemPtr& sys) { return sys; }
> Maybe this could just be a constructor of LCMCoordinateFrame instead? That would keep CoordinateFrame cleaner.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/RobotLocomotion/drake/pull/1274/files#r37701816>.
>